[AIRFLOW-1272] Google Cloud ML Batch Prediction Operator

jiwang576 opened the pull request:
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!


### JIRA
- [x] My PR addresses the following [AIRFLOW-1272](https://issues.apache.org/jira/browse/AIRFLOW-1272) issues and references them in the PR title. 


### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

- Added CloudMLBatchPredictionOperator, which is a wrapper around the Google Cloud ML Batch Prediction API.
- Augmented the CloudMLHook in PR#2379, which handles submitting job request to Google Cloud ML Engine.

### Tests
- [x] My PR adds the following unit tests:

- tests.contrib.operators.test_cloudml_operator:CloudMLBatchPredictionOperatorTest
- tests.contrib.hooks.test_gcp_cloudml_hook:TestCloudMLHook, which builds upon the same test class in PR#2379

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

Commit:
[AIRFLOW-1272] Add Google Cloud ML Batch Prediction Operator

There should be an Operator for the Google Cloud ML Batch Prediction
Job, which supports executing Tensorflow neural network prediction as a
service.
Documentation is here:
https://cloud.google.com/ml-engine/docs/how-tos/batch-predict.

Commit:
[AIRFLOW-1272] Add Google Cloud ML Batch Prediction Operator

There should be an Operator for the Google Cloud ML Batch Prediction
Job, which supports executing Tensorflow neural network prediction as a
service.
Documentation is here:
https://cloud.google.com/ml-engine/docs/how-tos/batch-predict.

saguziel commented on airflow/contrib/hooks/gcp_cloudml_hook.py:
```diff
@@ -0,0 +1,138 @@
+    """CloudML job operations helper class."""
+
+    def __init__(self, cloudml, project_name, job_id, job_spec=None):
+        assert project_name is not None and project_name is not ''
```
Should compared against '' with `!=` since logically we are checking equality, not identity. I'm not even sure that '' is guaranteed to be interned.

jiwang576 replied:
Thank you! The fix was committed.

jiwang576 committed:
[AIRFLOW-1272] Fix precondition in cloudml hook

Changed precondition assertions inside the gcp_cloudml_hook.py.

criccomini commented on airflow/utils/db.py:
```diff
@@ -120,6 +120,9 @@ def initdb():
             conn_id='bigquery_default', conn_type='bigquery'))
     merge_conn(
         models.Connection(
+            conn_id='google_cloud_default', conn_type='google_cloud_platform'))
```
I think this was already added as part of:

https://github.com/apache/incubator-airflow/pull/2379/files#diff-1fd56b45d5b414bd50b37baaac692c40R133

jiwang576 replied:
Yes these are duplicate code when PR2379 was not merged into Airflow. I will get rid of those.
Thank you!

criccomini replied:
Still needs to get cleaned up. Same declaration is here:

https://github.com/apache/incubator-airflow/pull/2390/files#diff-1fd56b45d5b414bd50b37baaac692c40R134

criccomini commented on airflow/utils/db.py:
```diff
@@ -130,6 +133,10 @@ def initdb():
             schema='hive', port=3400))
     merge_conn(
         models.Connection(
+            conn_id='google_cloud_default', conn_type='google_cloud_platform',
```
Why two google_cloud_default entries?

criccomini commented on airflow/contrib/hooks/gcp_cloudml_hook.py:
```diff
@@ -0,0 +1,139 @@
+
+    def create_job(self, project_name, job):
+        """
+        Creates a CloudML Job, and returns the Job object, which can be waited
```
> "which can be waited upon"

Based on the return line below `cloudml_job.wait_for_done(10)`, it looks to me like the waiting is done before `create_job` returns.

jiwang576 replied:
Thanks for the catch! We wrote a different docstring for that. Hopefully it reveals clearly what it does.

Commit:
[AIRFLOW-1272] Reorder helper functions

Commit:
[AIRFLOW-1272] Merge conflicts with PR#2379

Commit:
[AIRFLOW-1272] Merge remote-tracking branch

Commit:
[AIRFLOW-1272] Style-check merged files

criccomini commented on airflow/contrib/operators/cloudml_operator.py:
```diff
@@ -14,16 +14,321 @@
+    """
+
+    template_fields = [
+        '_model',
```
Probably want project_name templated, too

jiwang576 replied:
Thanks Chris! the project_name is not intended for templating on purpose.  

To answer the question of the funky rebase, I changed some of the code related to PR2379 for styles only. The main code changes are on the CloudMLBatchOperator (PR2390).

criccomini replied:
Got it, thanks for clarification.

criccomini reviewed (commented):
This rebase looks kind of funky. It's showing a bunch of #2379. Is everything as intended?

criccomini commented on airflow/contrib/operators/cloudml_operator.py:
```diff
@@ -14,16 +14,321 @@
+                job_id)
+        except errors.HttpError as e:
+            if e.resp.status == 404:
+                logging.error(
```
Is this really an error? Seems like `info` is more appropriate

jiwang576 replied:
Done. Thanks for the suggestion!

criccomini commented on airflow/contrib/operators/cloudml_operator.py:
```diff
@@ -14,16 +14,321 @@
+                logging.error(
+                    'Job with job_id {} does not exist. Will create it.'
+                    .format(job_id))
+                finished_prediction_job = hook.create_job(
```
Should wait_for_job_done be called here?

jiwang576 replied:
Sorry for the confusing interface. It's not needed as create_job is calling wait_for_job_done under the cover. 
To make it clearer we refactored the hook interface and how the operator uses it. Now only create_job (which creates and wait until a job is finished) is "exposed" and used by the operator.

criccomini reviewed (commented):
Overall, pretty good shape. Just a few nits/questions.

Commit:
[AIRFLOW-1272] Fix logging level and duplicated db setup

Commit:
[AIRFLOW-1272] Refactor hook code

This commit simplified the CloudMLHook class regarding job requests.
Public interface of the hook now only exposes the create_job API,
which contains the logic of creating the job, test for validation, and
polling for the job until it's finished.

Commit:
[AIRFLOW-1272] Apply cosmetic restructuring and commenting to hook

Commit:
[AIRFLOW-1272] Merge with upstream/master

Commit:
[AIRFLOW-1272] Remove trailing whitespace

Commit:
[AIRFLOW-1272] Fix typo in create_version comments
